channeldb/invoices.go

```diff
@@ -2225,6 +2231,83 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
 	return invoice, nil
 }
 
+// settleHodlInvoice marks a hodl invoice as settled.
+//
+// NOTE: Currently it is not possible to have HODL AMP invoices.
+func (d *DB) settleHodlInvoice(invoices, settleIndex kvdb.RwBucket,
+	invoiceNum []byte, invoice *invpkg.Invoice, hash *lntypes.Hash,
+	update *invpkg.InvoiceUpdateDesc) (*invpkg.Invoice, error) {
+
+	if invoice.IsAMP() {
+		return nil, fmt.Errorf("unable to settle hodl invoice: %v is "+
+			"an AMP invoice", invoice.AddIndex)
+	}
+
+	// TODO(positiveblue): because NewState can only be ContractSettled we
+	// can remove it from the API and set it here directly.
+	switch {
+	case update.State == nil:
+		fallthrough
+
+	case update.State.NewState != invpkg.ContractSettled:
+		fallthrough
+
+	case update.State.Preimage == nil:
+		return nil, fmt.Errorf("unable to settle hodl invoice: "+
```

Comment:
Maybe we could move this error to be returned in the above `case update.State.NewState != invpkg.ContractSettled:` and instead return an error to indicate the `Preimage` is nil, just thinking of being explicit.